rust/ql/src/queries/unusedentities/UnreachableCode.ql

```diff
@@ -0,0 +1,59 @@
+/**
+ * @name Unreachable code
+ * @description Code that cannot be reached should be deleted.
+ * @kind problem
+ * @problem.severity recommendation
+ * @precision medium
+ * @id rust/dead-code
+ * @tags maintainability
+ */
+
+import rust
+import codeql.rust.controlflow.ControlFlowGraph
+import codeql.rust.controlflow.internal.ControlFlowGraphImpl as ControlFlowGraphImpl
+
+/**
+ * Holds if `n` is an AST node that's unreachable, and is not the successor
+ * of an unreachable node (which would be a duplicate result).
+ */
+predicate firstUnreachable(AstNode n) {
```

Comment:
I think it would be better to base this on the resulting CFG, as that can take things like splitting into account (if we implement that). For example, it would be able to identify
```
if (x) {
  ...
  if (!x) {
   // dead
  }
}
```

The CFG construction only assigns CFG nodes to AST nodes that are actually reachable, so it would be something like
```
private predicate unreachable(AstNode n) {
  not n = any(CfgNode cfn).getAstNode()
)

private predicate firstUnreachable(AstNode n) {
  unreachable(n) and
  (
    not ControlFlowGraphImpl::succ(_, n, _)
    or
    exists(AstNode pred |
      ControlFlowGraphImpl::succ(pred, n, _) and
      not unreachable(pred)
    )
  )
)
```

Reply:
I've added your example and switched to your implementation of `firstUnreachable` (though at the moment results are identical).